numpy/core/src/multiarray/array_assign_array.c

```diff
@@ -345,6 +345,21 @@ PyArray_AssignArray(PyArrayObject *dst, PyArrayObject *src,
         }
     }
 
+    /* optimization: scalar boolean mask */
+    if (wheremask != NULL &&
+            PyArray_NDIM(wheremask) == 0 &&
+            PyArray_DESCR(wheremask)->type_num == NPY_BOOL) {
```

Comment:
according to the function comment  this check might not be needed, since wheremask must be a bool array.

I'd keep it though, it doesn't hurt.